src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -375,5 +375,107 @@ private static bool IsNullable(Type type)
+                   .SingleOrDefault(candidate =>
+                   {
+                       if (candidate.Name != name || candidate.GetGenericArguments().Length != 0) return false;
+
+                       var parameters = candidate.GetParameters();
+                       if (parameters.Length != parameterTypes.Length) return false;
+
+                       for (var i = 0; i < parameterTypes.Length; i++)
+                           if (parameters[i].ParameterType != parameterTypes[i])
+                               return false;
+
+                       return true;
+                   });
+
+                if (method != null) return method;
+            }
+
+            return null;
+        }
+
+        internal static PropertyInfo GetPublicInstanceProperty(this Type type, string name, Type[] indexParameterTypes)
+        {
+            for (var currentType = type; currentType != null; currentType = currentType.GetTypeInfo().BaseType)
+            {
+                var property = currentType
+                     .GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
+                     .SingleOrDefault(candidate =>
+                     {
+                         if (candidate.Name != name) return false;
+
+                         var indexParameters = candidate.GetIndexParameters();
+                         if (indexParameters.Length != indexParameterTypes.Length) return false;
+
+                         for (var i = 0; i < indexParameterTypes.Length; i++)
+                             if (indexParameters[i].ParameterType != indexParameterTypes[i]) return false;
+
+                         return true;
+                     });
+
+                if (property != null) return property;
+            }
+
+            return null;
+        }
+
+        internal static object InvokeWithTransparentExceptions(this MethodBase methodBase, object instance)
+        {
+            // If we ever target .NET Core 2.1, we can keep from mucking with the exception stack trace
+            // using BindingFlags.DoNotWrapExceptions rather than try…catch.
+
+            try
+            {
+                return methodBase.Invoke(instance, null);
+            }
+            catch (TargetInvocationException ex)
+            {
+                ExceptionHelper.Rethrow(ex.InnerException);
+                throw null; // Rethrow’s return type would be `never` if C# could express that.
+            }
+        }
+
+        internal static object DynamicInvokeWithTransparentExceptions(this Delegate @delegate)
+        {
+            try
+            {
+                return @delegate.DynamicInvoke();
+            }
+            catch (TargetInvocationException ex)
+            {
+                ExceptionHelper.Rethrow(ex.InnerException);
+                throw null; // Rethrow’s return type would be `never` if C# could express that.
```

Comment:
Just checking if I understand this, the `throw null` causes the `RecordException` to return null if there is no inner exception, correct?

Reply:
`throw null;` is there to satisfy C#'s flow control but it will never be hit unless `ExceptionHelper.Rethrow` fails to throw an exception. (I think its actual effect is identical to `throw new NullReferenceException();`.)

Maybe better would be:

```cs
throw new InvalidOperationException("ExceptionHelper.Rethrow failed to throw an exception.");
```